edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/Handlers/ConnectionHandler.cs

```diff
@@ -0,0 +1,294 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Text;
+    using System.Threading.Channels;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Edge.Hub.Core;
+    using Microsoft.Azure.Devices.Edge.Hub.Core.Device;
+    using Microsoft.Azure.Devices.Edge.Hub.Core.Identity;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Azure.Devices.Edge.Util.Concurrency;
+    using Microsoft.Extensions.Logging;
+    using Newtonsoft.Json;
+
+    public class ConnectionHandler : IConnectionRegistry, ISubscriber, IMessageConsumer
+    {
+        const string TopicDeviceConnected = "$edgehub/connected";
+
+        static readonly string[] subscriptions = new[] { TopicDeviceConnected };
+
+        readonly IConnectionProvider connectionProvider;
+        readonly IIdentityProvider identityProvider;
+        readonly ISystemComponentIdProvider systemComponentIdProvider;
+        readonly DeviceProxy.Factory deviceProxyFactory;
+
+        readonly Channel<MqttPublishInfo> notifications;
+        readonly Task processingLoop;
+
+        AsyncLock guard = new AsyncLock();
+
+        // Normal dictionary would be sufficient because of the locks, however we need AddOrUpdate()
+        ConcurrentDictionary<IIdentity, IDeviceListener> knownConnections = new ConcurrentDictionary<IIdentity, IDeviceListener>();
+
+        // this class is auto-registered so no way to implement an async activator.
+        // hence this one needs to get a Task<T> which is suboptimal, but that is the way
+        // IConnectionProvider is registered
+        public ConnectionHandler(Task<IConnectionProvider> connectionProvider, IIdentityProvider identityProvider, ISystemComponentIdProvider systemComponentIdProvider, DeviceProxy.Factory deviceProxyFactory)
+        {
+            if (!connectionProvider.IsCompleted)
+            {
+                // if this leads to a dead-lock, at least it gets logged.
+                Events.BlockingDependencyInjection();
+            }
+
+            this.connectionProvider = connectionProvider.Result;
```

Comment:
Pls don't do this.. we use only async/await everywhere in the code as a principle.

Reply:
fixed by storing the Task<x> and using with 'await' when connectionProvider is necessary